### PR TITLE
Stats no vector

### PR DESCRIPTION
**Pull request by pradeepcrao**
Improve cpu and memory usage of the sink for counters, gauges and text readouts by:

Iterating over stats in the store to create a snapshot (instead of
creating a vector by iterating over scopes.)

Commit Message:
Additional Description:
Risk Level: Low
Testing: Added benchmark test for stats sink.
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Benchmark test results of server_stats_flush_benchmark_test:

With change:
name                     cpu/op
bmFlushToSinks/10         778ns ± 1%
bmFlushToSinks/100       2.83µs ± 3%
bmFlushToSinks/1000      39.5µs ± 1%
bmFlushToSinks/10000      409µs ± 4%
bmFlushToSinks/100000    6.07ms ±19%
bmFlushToSinks/1000000    100ms ± 4%

Without change:
name                     cpu/op
bmFlushToSinks/10        4.44µs ± 4%
bmFlushToSinks/100       31.4µs ± 2%
bmFlushToSinks/1000       376µs ± 2%
bmFlushToSinks/10000     5.40ms ± 7%
bmFlushToSinks/100000    90.1ms ± 4%
bmFlushToSinks/1000000    1.59s ± 4%

### CODE REVIEW

File: envoy/stats/allocator.h [hunk 1]
```diff
@@ -58,6 +58,20 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Iterate over all stats that need to be sinked. Note, that implementations can potentially hold
+   * on to a mutex that will deadlock if the passed in functors try to create or delete a stat.
+   * @param f_size functor that is provided the number of all sinked stats. Note this is called
+   *        only once.
```
Comment (jmarantz) [inline envoy/stats/allocator.h:65]: s/once./once, prior to any calls to f_stat/

File: envoy/stats/allocator.h [hunk 2]
```diff
@@ -58,6 +58,20 @@ class Allocator {
...
+   * @param f_stat functor that is provided one sinked stat at a time.
+   */
+  virtual void forEachSinkedCounter(std::function<void(std::size_t)> f_size,
+                                    std::function<void(Stats::Counter&)> f_stat) PURE;
+  virtual void forEachSinkedGauge(std::function<void(std::size_t)> f_size,
+                                  std::function<void(Stats::Gauge&)> f_stat) PURE;
+  virtual void forEachSinkedTextReadout(std::function<void(std::size_t)> f_size,
+                                        std::function<void(Stats::TextReadout&)> f_stat) PURE;
```
Comment (jmarantz) [inline envoy/stats/allocator.h:73]: it might be convenient for callers to receive these as `const Stats::CounterSharedPtr&` etc rather than `Stats::Counter&`. That way if a caller decided it wanted to collect them and sort them, it could. I admit that would defeat the purpose of the new API, but it would also mean you could implement the old API on top of the new one, and remove the old API (potentially).

It would also make it easier to (for example) have an admin option to do some alternative sorting, or doing a custom sort after doing its own regex filtering.

And I don't think you'd lose any performance or generality.

File: test/server/server_stats_flush_benchmark_test.cc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  Stats::SymbolTableImpl symbol_table_;
+  Stats::StatNamePool pool_;
+  Stats::AllocatorImpl stats_allocator_;
+  Stats::ThreadLocalStoreImpl stats_store_;
+  std::list<Stats::SinkPtr> sinks_;
+  Event::SimulatedTimeSystem time_system_;
+};
+
+static void bmLarge(benchmark::State& state) {
+  uint64_t n_counters = 1000000, n_gauges = 1000000, n_text_readouts = 1000000;
```
Comment (jmarantz) [inline test/server/server_stats_flush_benchmark_test.cc:60]: s/n_counters/num_counters/ here and elsewhere in the PR

Actually it'd be better to parameterize this test on this number. Look at other "*_speed_test.cc" files for examples. Then you'd need only one test method, and you could parameterize a wider spectrum of values, so you could see where the knee of the curve was more easily than having just 2 data points.

File: test/server/server_stats_flush_benchmark_test.cc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  StatsSinkFlushSpeedTest speed_test(n_counters, n_gauges, n_text_readouts);
+  speed_test.test(state);
+}
+BENCHMARK(bmLarge)->Unit(::benchmark::kMillisecond);
+
+static void bmSmall(benchmark::State& state) {
```
Comment (jmarantz) [inline test/server/server_stats_flush_benchmark_test.cc:66]: Most tests of this sort are called "BM_xxx" which is some benchmarking convention, but violates Envoy style guide, requiring clang-tidy "nolint" comments. What you have might be better. We'll see if anything else has a problem with this.

File: test/server/server_test.cc [hunk 5]
```diff
@@ -101,6 +101,57 @@ TEST(ServerInstanceUtil, flushHelper) {
   InstanceUtil::flushMetricsToSinks(sinks, mock_store, time_system);
 }
 
+TEST(ServerInstanceUtil, sinkFilter) {
```
Comment (jmarantz) [inline test/server/server_test.cc:104]: SinkFilter

File: test/server/server_test.cc [hunk 6]
```diff
@@ -101,6 +101,57 @@ TEST(ServerInstanceUtil, flushHelper) {
...
+  InSequence s;
+
+  Stats::TestUtil::TestStore store;
+  Event::SimulatedTimeSystem time_system;
+  // Set the sink filters
+  store.setCounterSinkFilter(
+      [](const Stats::Counter& counter) { return counter.name().compare("hello") == 0; });
```
Comment (jmarantz) [inline test/server/server_test.cc:111]: `counter.name() == "hello"` here & below

File: test/server/server_test.cc [hunk 7]
```diff
@@ -101,6 +101,57 @@ TEST(ServerInstanceUtil, flushHelper) {
...
+  std::list<Stats::SinkPtr> sinks;
+  InstanceUtil::flushMetricsToSinks(sinks, store, time_system);
+  // Make sure that counters have been latched even if there are no sinks.
+  EXPECT_EQ(2UL, c.value());
+  EXPECT_EQ(0, c.latch());
+
+  Stats::MockSink* sink = new StrictMock<Stats::MockSink>();
+  sinks.emplace_back(sink);
+  EXPECT_CALL(*sink, flush(_)).WillOnce(Invoke([](Stats::MetricSnapshot& snapshot) {
+    ASSERT_EQ(snapshot.counters().size(), 1);
```
Comment (jmarantz) [inline test/server/server_test.cc:138]: capture snapshot.counters() in a temp rather than repeating the code.

File: test/server/server_stats_flush_benchmark_test.cc [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    for (uint64_t idx = 0; idx < n_text_readouts; ++idx) {
+      auto stat_name = pool_.add(absl::StrCat("text_readout.", idx));
+      stats_store_.textReadoutFromStatName(stat_name).set(absl::StrCat("text_readout.", idx));
+    }
+  }
+
+  void test(benchmark::State& state) {
+    for (auto _ : state) {
+      UNREFERENCED_PARAMETER(state);
+      Server::InstanceUtil::flushMetricsToSinks(sinks_, stats_store_, time_system_);
```
Comment (jmarantz) [inline test/server/server_stats_flush_benchmark_test.cc:46]: Question: does this test actually prove that things will be faster/smaller when running with a live sink?

I noticed that MetricSnapshot's API still requires returning a vector of all counters. Have we just moved the construction of that vector but not eliminated it?

File: source/server/server.cc [hunk 9]
```diff
@@ -165,30 +166,26 @@ void InstanceImpl::failHealthcheck(bool fail) {
...
 
 MetricSnapshotImpl::MetricSnapshotImpl(Stats::Store& store, TimeSource& time_source) {
-  snapped_counters_ = store.counters();
-  counters_.reserve(snapped_counters_.size());
-  for (const auto& counter : snapped_counters_) {
-    counters_.push_back({counter->latch(), *counter});
-  }
+  store.forEachSinkedCounter([this](std::size_t size) mutable { counters_.reserve(size); },
+                             [this](Stats::Counter& counter) mutable {
+                               counters_.push_back({counter.latch(), counter});
```
Comment (jmarantz) [inline source/server/server.cc:171]: I think we need to plumb this change more deeply to actually provide a benefit in practice. I would argue that makes sense in this PR. WDYT?

File: envoy/stats/allocator.h (hunk 2 above)
Comment (pradeepcrao) [reply envoy/stats/allocator.h:73]: That would require creating the shared pointer from raw pointers and passing it to the functor. Would it not be better for the functor to do that (and pay the cost) if it needs to?

File: test/server/server_stats_flush_benchmark_test.cc (hunk 8 above)
Comment (pradeepcrao) [reply test/server/server_stats_flush_benchmark_test.cc:46]: The snapshotting into vectors happens within flushMetricsToSinks, so this test should capture the impact of that.

File: envoy/stats/allocator.h (hunk 2 above)
Comment (jmarantz) [reply envoy/stats/allocator.h:73]: Yes you are right. I was thinking you had the RefcountPtr (not std::shared_ptr) held in the set you were iterating over, but actually it's just got raw pointers.

And because these are holding the ref-count in a member variable you can actually construct RefcountPtr from a raw ptr if needed:

https://github.com/envoyproxy/envoy/blob/74d5ef8bb36bfcbdd89555a4ea020519033699f5/envoy/stats/refcount_ptr.h#L35

File: test/server/server_stats_flush_benchmark_test.cc (hunk 8 above)
Comment (jmarantz) [reply test/server/server_stats_flush_benchmark_test.cc:46]: ok, sgtm.

File: test/server/server_stats_flush_benchmark_test.cc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  std::list<Stats::SinkPtr> sinks_;
+  Event::SimulatedTimeSystem time_system_;
+};
+
+static void bmFlushToSinks(benchmark::State& state) {
+  StatsSinkFlushSpeedTest speed_test(state.range(0));
+  speed_test.test(state);
+}
+BENCHMARK(bmFlushToSinks)->Unit(::benchmark::kMillisecond)->Range(10, 100000);
+
```
Comment (jmarantz) [inline test/server/server_stats_flush_benchmark_test.cc:64]: maybe add a param of 1000?

File: test/server/server_test.cc [hunk 11]
```diff
@@ -679,6 +679,17 @@ TEST_P(ServerInstanceImplTest, ShutdownBeforeWorkersStarted) {
...
+  options_.service_cluster_name_ = "some_cluster_name";
+  options_.service_node_name_ = "some_node_name";
+  try {
+    initialize("test/server/test_data/server/unparseable_bootstrap.yaml");
+    FAIL();
+  } catch (const EnvoyException& e) {
+    EXPECT_THAT(e.what(), HasSubstr("Unable to parse JSON as proto"));
+  }
+}
+
```
Comment (jmarantz) [inline test/server/server_test.cc:692]: is it possible some unrelated stuff has comingled with this PR?

Comment (pradeepcrao) [reply test/server/server_test.cc]: fixed.

File: test/server/server_stats_flush_benchmark_test.cc (hunk 10 above)
Comment (pradeepcrao) [reply test/server/server_stats_flush_benchmark_test.cc:64]: Added RangeMultiplier of 10

File: source/server/server.cc (hunk 9 above)
Comment (pradeepcrao) [reply source/server/server.cc:171]: I am not sure I follow :-(  Could you please elaborate?

Comment (jmarantz) [reply source/server/server.cc:171]: I think you're all set based on our IM conversation. resolving.

File: test/server/server_stats_flush_benchmark_test.cc [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+#include "test/test_common/utility.h"
+
+#include "absl/strings/str_cat.h"
+#include "benchmark/benchmark.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+class StatsSinkFlushSpeedTest {
+
```
Comment (jmarantz) [inline test/server/server_stats_flush_benchmark_test.cc:22]: nit: remove this blank line and put it between the `class` and `namespace` lines.

File: test/server/server_stats_flush_benchmark_test.cc (hunk 8 above)
Comment (jmarantz) [reply test/server/server_stats_flush_benchmark_test.cc:49]: should we clear sinks_ on each iter, or just declare and init the sinks inside the loop?

File: test/server/server_stats_flush_benchmark_test.cc (hunk 10 above)
Comment (jmarantz) [reply test/server/server_stats_flush_benchmark_test.cc:73]: nice, thanks.

**Review by jmarantz — COMMENTED** [review]
This basically looks right to me, but I think @mattklein123 needs to weigh in. It so happens that on another PR discussion we were thinking we might want to use info we propose to store in the Scope during the sink process for Prometheus.

**Review by mattklein123 — DISMISSED** [review]
This LGTM modulo the remaining @jmarantz comments. I think if we decide to store something in the scope we can probably figure out a way to do that as a future change? (Since right now we don't provide scope information to the sinks as it is.)

File: test/server/server_stats_flush_benchmark_test.cc (hunk 8 above)
Comment (pradeepcrao) [reply test/server/server_stats_flush_benchmark_test.cc]: Declared sinks inside the loop.

File: source/common/stats/allocator_impl.cc [hunk 13]
```diff
@@ -38,6 +25,23 @@ const char AllocatorImpl::DecrementToZeroSyncPoint[] = "decrement-zero";
 AllocatorImpl::~AllocatorImpl() {
   ASSERT(counters_.empty());
   ASSERT(gauges_.empty());
+
+  // Move deleted stats into the sets for removeFromSetLockHeld to function.
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:29]: s/for removeSetLockHeld/for the ASSERTs in removeFromSetLockHeld/

I think you could actually surround all this in `ifndef NDEBUG` and not go through these loops at all in production.

File: source/common/stats/allocator_impl.cc [hunk 14]
```diff
@@ -38,6 +25,23 @@ const char AllocatorImpl::DecrementToZeroSyncPoint[] = "decrement-zero";
...
+  for (auto& counter : deleted_counters_) {
+    // Assert that there were no duplicates.
+    ASSERT(counters_.count(counter.get()) == 0);
+    counters_.insert(counter.get());
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:33]: You can do this in one statement with `ASSERT(counters.insert(counter.get()).second)`, or in two with
```
auto insertion = counters.insert(counter.get());
ASSERT(insertion.second);
```
The latter feels more like it more closely mirrors the pattern in removeFromSetLockHeld, and is just as fast (only does one map operation).

File: source/common/stats/allocator_impl.cc [hunk 15]
```diff
@@ -362,7 +366,7 @@ void AllocatorImpl::markCounterForDeletion(StatName name) {
   if (iter == counters_.end()) {
     return;
   }
-  ASSERT(!hasStat(deleted_counters_, *iter));
+  // Duplicates are checked in ~AllocatorImpl.
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:369]: s/checked/ASSERTed/ to make it clearer that doesn't happen in production.

**Review by jmarantz — COMMENTED** [review]
looks great modulo minor nits!

**Comment by jmarantz** [comment]
You still have to change expectations in a the scope deleter test, right?

File: source/common/stats/allocator_impl.cc (hunk 13 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:29]: Done

File: source/common/stats/allocator_impl.cc (hunk 14 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:33]: Done

File: source/common/stats/allocator_impl.cc (hunk 15 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:369]: Done

**Comment by pradeepcrao** [comment]
> You still have to change expectations in a the scope deleter test, right?

Are you referring to thread_local_store_test.cc ScopeDelete? I fixed that.

File: source/common/stats/allocator_impl.cc (hunk 14 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:33]: Done

File: test/common/stats/thread_local_store_test.cc [hunk 16]
```diff
@@ -508,10 +508,11 @@ TEST_F(StatsThreadLocalStoreTest, ScopeDelete) {
   EXPECT_CALL(main_thread_dispatcher_, post(_));
   EXPECT_CALL(tls_, runOnAllThreads(_, _)).Times(testing::AtLeast(1));
   scope1.reset();
-  EXPECT_EQ(0UL, store_->counters().size());
+  EXPECT_EQ(1UL, store_->counters().size());
```
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:511]: just add a comment here that the counter is gone from all scopes, but is still held by the local variable so it won't be removed from the allocator or store.

File: test/server/server_stats_flush_benchmark_test.cc (hunk 12 above)
Comment (jmarantz) [reply test/server/server_stats_flush_benchmark_test.cc:23]: now you added the blank line below `namespace` but there's still a superfluous one below `class`.

File: source/common/stats/allocator_impl.cc [hunk 17]
```diff
@@ -316,5 +335,65 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
   return new CounterImpl(name, *this, tag_extracted_name, stat_name_tags);
 }
 
+void AllocatorImpl::forEachCounter(std::function<void(std::size_t)> f_size,
```
[the remark this answers is not in the document]
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:338]: I had them in the other PR with the filtering. I'll copy them over here without the filtering.

File: tools/spelling/spelling_dictionary.txt [hunk 18]
```diff
@@ -15,6 +15,7 @@ ARN
 ASAN
 ASCII
 ASM
+ASSER
```
[the remark this answers is not in the document]
Comment (pradeepcrao) [reply tools/spelling/spelling_dictionary.txt:18]: changed to ASSERTed

File: envoy/stats/allocator.h [hunk 19]
```diff
@@ -58,6 +58,28 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Mark rejected stats as deleted by moving them to a different vector, so they don't show up
+   * when iterating over stats, but prevent crashes when trying to accesses references to them.
+   */
+  virtual void markCounterForDeletion(StatName name) PURE;
```
[the remark this answers is not in the document]
Comment (pradeepcrao) [reply envoy/stats/allocator.h:65]: The condition is still needed as stat names are duplicated across scopes, and we iterate over scopes to mark for deletion. However, I agree it can be jarring to see markCounterForDeletion take a StatName and not a counter.
Making the change.

File: test/common/stats/thread_local_store_test.cc (hunk 16 above)
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:511]: Done.

File: source/common/stats/allocator_impl.cc [hunk 20]
```diff
@@ -316,5 +336,79 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
...
+void AllocatorImpl::forEachGauge(std::function<void(std::size_t)> f_size,
+                                 std::function<void(Stats::Gauge&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  size_t num_gauges = 0;
+  std::for_each(gauges_.begin(), gauges_.end(), [&num_gauges](Gauge* const gauge) mutable {
+    if (gauge->importMode() != Gauge::ImportMode::Uninitialized) {
+      num_gauges++;
+    }
+  });
+  f_size(num_gauges);
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:357]: sorry a few more nits. Can you skip this loop if f_size is nullptr?

File: source/common/stats/allocator_impl.cc [hunk 21]
```diff
@@ -316,5 +336,79 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
...
+  for (auto& counter : counters_) {
+    f_stat(*counter);
+  }
+}
+
+void AllocatorImpl::forEachGauge(std::function<void(std::size_t)> f_size,
+                                 std::function<void(Stats::Gauge&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  size_t num_gauges = 0;
+  std::for_each(gauges_.begin(), gauges_.end(), [&num_gauges](Gauge* const gauge) mutable {
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:352]: std::count_if

File: source/common/stats/allocator_impl.cc [hunk 22]
```diff
@@ -316,5 +336,79 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
...
+                                   std::function<void(Stats::Counter&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  f_size(counters_.size());
+  for (auto& counter : counters_) {
+    f_stat(*counter);
+  }
+}
+
+void AllocatorImpl::forEachGauge(std::function<void(std::size_t)> f_size,
+                                 std::function<void(Stats::Gauge&)> f_stat) const {
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:349]: it's not super-obvious that everyone who would want to call forEachGauge on an allocator would want to skip the uninitialized ones.

I'm wondering if it makes more sense to just let the caller do this filtering. The main argument for having this be a feature of this API is that you could hold the lock while counting them and then enumerating them.

Maybe let the caller provide the predicate, and offer that as a feature for the other stat types?

I'm OK if you want to do that as a follow-up, but it'd be nice to offer a consistent API eventually.

File: source/common/stats/allocator_impl.h [hunk 23]
```diff
@@ -11,6 +11,7 @@
 
 #include "absl/container/flat_hash_set.h"
 #include "absl/strings/string_view.h"
+#include "metric_impl.h"
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.h:14]: `#include "source/common/stats/metric_impl.h"`

File: test/common/stats/allocator_impl_test.cc [hunk 24]
```diff
@@ -25,12 +27,18 @@ class AllocatorImplTest : public testing::Test {
 
   void clearStorage() {
     pool_.clear();
-    EXPECT_EQ(0, symbol_table_.numSymbols());
+    // If stats have been marked for deletion, they are not cleared until the
+    // destructor of alloc_ is called, and hence the symbol_table_.numSymbols()
+    // will be greater than zero at this point.
+    if (are_stats_marked_for_deletion_ == false) {
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:33]: `if (!are_stats_marked_for_deletion_)` ...

File: test/common/stats/allocator_impl_test.cc [hunk 25]
```diff
@@ -125,6 +133,177 @@ TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
   EXPECT_FALSE(alloc_.isMutexLockedForTest());
 }
 
+TEST_F(AllocatorImplTest, ForEachCounter) {
+
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:137]: rm blank line

File: test/common/stats/allocator_impl_test.cc [hunk 26]
```diff
@@ -125,6 +133,177 @@ TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
...
+  counters.clear();
+  num_iterations = 0;
+  alloc_.forEachCounter([&num_counters](std::size_t size) { num_counters = size; },
+                        [&num_iterations](Stats::Counter&) { ++num_iterations; });
+  EXPECT_EQ(num_counters, 0);
+  EXPECT_EQ(num_iterations, 0);
+}
+
+TEST_F(AllocatorImplTest, ForEachGauge) {
+
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:191]: rm blank line

File: test/common/stats/allocator_impl_test.cc [hunk 27]
```diff
@@ -125,6 +133,177 @@ TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
...
+  gauges.clear();
+  num_iterations = 0;
+  alloc_.forEachGauge([&num_gauges](std::size_t size) { num_gauges = size; },
+                      [&num_iterations](Stats::Gauge&) { ++num_iterations; });
+  EXPECT_EQ(num_gauges, 0);
+  EXPECT_EQ(num_iterations, 0);
+}
+
+TEST_F(AllocatorImplTest, ForEachTextReadout) {
+
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:253]: rm blank line

File: test/common/stats/thread_local_store_test.cc [hunk 28]
```diff
@@ -508,10 +508,15 @@ TEST_F(StatsThreadLocalStoreTest, ScopeDelete) {
   EXPECT_CALL(main_thread_dispatcher_, post(_));
   EXPECT_CALL(tls_, runOnAllThreads(_, _)).Times(testing::AtLeast(1));
   scope1.reset();
-  EXPECT_EQ(0UL, store_->counters().size());
+  // The counter is gone from all scopes, but is still held in the local
+  // variable c1. Hence, it will no be removed from the allocator or store.
```
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:512]: s/no/not/ (typo)

**Review by jmarantz — COMMENTED** [review]
really great stuff. Just a few more nits and some thoughts about API consistency.

File: source/common/stats/allocator_impl.cc (hunk 22 above)
Comment (jmarantz) [reply source/common/stats/allocator_impl.cc:349]: I could see you might want to similarly iterated over counters where used()==true

File: source/common/stats/allocator_impl.cc (hunk 20 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:357]: Done

File: source/common/stats/allocator_impl.cc (hunk 21 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:352]: Outdated, per new changes.

File: source/common/stats/allocator_impl.cc (hunk 22 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:349]: Made this change.

Added the check to gauges().

File: source/common/stats/allocator_impl.h (hunk 23 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.h:14]: Not sure how that got added. Done.

File: test/common/stats/allocator_impl_test.cc (hunk 24 above)
Comment (pradeepcrao) [reply test/common/stats/allocator_impl_test.cc:33]: Done

File: test/common/stats/allocator_impl_test.cc (hunk 27 above)
Comment (pradeepcrao) [reply test/common/stats/allocator_impl_test.cc:253]: Done

File: test/common/stats/thread_local_store_test.cc (hunk 28 above)
Comment (pradeepcrao) [reply test/common/stats/thread_local_store_test.cc:512]: Done

**Review by jmarantz — COMMENTED** [review]
something went wrong in CI (`check linux_x64 gcc`) and I see this failure in your logs:
```
2021-09-02T00:45:00.5279225Z [ RUN      ] AllocatorImplTest.ForEachCounter
2021-09-02T00:45:00.5279626Z pure virtual method called
2021-09-02T00:45:00.5280002Z terminate called without an active exception
```

Not sure why this happens in that compiler variant. I'll let you sort it out :)

**Comment by pradeepcrao** [comment]
> something went wrong in CI (`check linux_x64 gcc`) and I see this failure in your logs:
> 
> ```
> 2021-09-02T00:45:00.5279225Z [ RUN      ] AllocatorImplTest.ForEachCounter
> 2021-09-02T00:45:00.5279626Z pure virtual method called
> 2021-09-02T00:45:00.5280002Z terminate called without an active exception
> ```
> 
> Not sure why this happens in that compiler variant. I'll let you sort it out :)

Fixed, and added comment wrt crash.

File: source/common/stats/allocator_impl.cc [hunk 29]
```diff
@@ -25,6 +26,29 @@ const char AllocatorImpl::DecrementToZeroSyncPoint[] = "decrement-zero";
...
+    ASSERT(insertion.second);
+  }
+  for (auto& text_readout : deleted_text_readouts_) {
+    auto insertion = text_readouts_.insert(text_readout.get());
+    // Assert that there were no duplicates.
+    ASSERT(insertion.second);
+  }
+  // Make sure all stats are deleted BEFORE threads are shutdown. This is
+  // because the stat destructors need to lock a mutex, which will lead to
+  // undefined behavior once the threads are destroyed.
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:48]: ah I see. Your comment is not quite accurate though, and there's a slightly nicer fix.

The problem is order-of-destruction and I think a better fix is to move the mutex to be above the declarations for all the collections in the structure. Then you don't need these explicit clear()s.

File: test/common/stats/allocator_impl_test.cc [hunk 30]
```diff
@@ -125,6 +133,166 @@ TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
   EXPECT_FALSE(alloc_.isMutexLockedForTest());
 }
 
+TEST_F(AllocatorImplTest, ForEachCounter) {
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:136]: can you add another test where you supply nullptr for the lambdas to each foreach? There'll be not much to check; just want to avoid crashing.

File: source/common/stats/allocator_impl.cc (hunk 29 above)
Comment (pradeepcrao) [reply source/common/stats/allocator_impl.cc:48]: I don't know how I missed something that simple :-/

Fixed.

File: test/common/stats/allocator_impl_test.cc (hunk 30 above)
Comment (pradeepcrao) [reply test/common/stats/allocator_impl_test.cc:136]: Done.

**Review by jmarantz — DISMISSED** [review]
Great job! Silently makes everything faster.

@mattklein123 do you want to look any further? Or just merge?

**Comment by mattklein123** [comment]
> @mattklein123 do you want to look any further? Or just merge?

Let me take a pass through tomorrow if that is OK?

**Comment by jmarantz** [comment]
Tomorrow would be great!

File: source/common/stats/allocator_impl.cc [hunk 31]
```diff
@@ -316,5 +336,77 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
...
+  Thread::LockGuard lock(mutex_);
+  auto iter = counters_.find(counter->statName());
+  if (iter == counters_.end()) {
+    // This has already been marked for deletion.
+    return;
+  }
+  ASSERT(counter.get() == *iter);
+  // Duplicates are ASSERTed in ~AllocatorImpl.
+  deleted_counters_.emplace_back(*iter);
+  counters_.erase(iter);
```
Comment (jmarantz) [inline source/common/stats/allocator_impl.cc:382]: I thought of a concern with this tactic for this class: if someone allocates a counter with the same name as one that was marked for deletion, we'll wind with two different counter objects with the same name. That might be OK but we might want to add a test.

It won't actually happen with ThreadLocalStore because it only calls markCounterForDeletion on stats that are rejected by the matcher, and the matcher can't be changed once it's added. The reason that this is needed is that the matcher can be added after some stats are created, and we need to effectively get rid of the now-rejected stats without causing references to freed memory.

Possible remedies I thought of include:
 * add a test and accept that as a weird use-case for this class, that won't be hit by ThreadLocalStore.
 * prevent this by checking against deleted_counters_ (which could be a set) during allocation
 * declare this invalid by asserting against deleted_counters_ for debug builds
 * use a bit in `flags_` to denote that a stat needs be considered as marked for deletion, and skip that during forEach. You could report the correct size by keeping track of the number of deleted counters in the class. I think this is the biggest change from what you have, but would mean that if you try to allocate a stat with the same name as one that was deleted, you'll get back the previously deleted stat.

File: envoy/stats/allocator.h [hunk 32]
```diff
@@ -58,6 +58,28 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Mark rejected stats as deleted by moving them to a different vector, so they don't show up
+   * when iterating over stats, but prevent crashes when trying to accesses references to them.
```
Comment (mattklein123) [inline envoy/stats/allocator.h:63]:
```suggestion
   * when iterating over stats, but prevent crashes when trying to access references to them.
```

File: source/common/stats/allocator_impl.cc (hunk 31 above)
Comment (mattklein123) [reply source/common/stats/allocator_impl.cc:382]: > use a bit in flags_ to denote that a stat needs be considered as marked for deletion, and skip that during forEach. You could report the correct size by keeping track of the number of deleted counters in the class. I think this is the biggest change from what you have, but would mean that if you try to allocate a stat with the same name as one that was deleted, you'll get back the previously deleted stat.

This approach sounds good to me, fwiw.

Stepping back though, I'm trying to understand why this logic was moved into the allocator versus where it used to be in thread local store? Can you help me understand that?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with one typo and one question, thanks.

/wait

**Comment by pradeepcrao** [comment]
Hi Matt,

With this change we iterate over stats in the Allocator instead of scopes as this is much faster. However, the Allocator did not know anything about rejected stats. So, we ended up including rejected stats when iterating over all stats.

To avoid iterating over them, I removed these stats from the StatSets in the Allocator. There is logic that looks for the stat in the Allocator StatSet when it's refcount goes to zero. To manage this cleanly, I moved the deleted stats from the Store to the Allocator.

Did that answer your question?

With regards to the issue that Josh mentioned, we decided that his first suggestion might be the best option, given that adding the deleted bit adds a lot of complexity and is only needed for an edge case that currently can't be hit in Envoy.
Does that sound acceptable to you?

**Comment by mattklein123** [comment]
> Did that answer your question?

Yup, thanks.

> Does that sound acceptable to you?

Sure, sounds good.

File: envoy/stats/allocator.h (hunk 32 above)
Comment (pradeepcrao) [reply envoy/stats/allocator.h]: Done.

File: envoy/stats/allocator.h [hunk 33]
```diff
@@ -58,6 +58,28 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Mark rejected stats as deleted by moving them to a different vector, so they don't show up
+   * when iterating over stats, but prevent crashes when trying to access references to them.
```
Comment (jmarantz) [inline envoy/stats/allocator.h:63]: Note here the surprising behavior that if you allocate a stat after having done this, you'll get a new one, and that callers should seek to avoid this situation (as ThreadLocalStore does).

File: test/common/stats/thread_local_store_test.cc [hunk 34]
```diff
@@ -1192,6 +1197,53 @@ TEST_F(StatsThreadLocalStoreTest, RemoveRejectedStats) {
...
+  EXPECT_EQ(0, store_->gauges().size());
+  EXPECT_EQ(0, store_->textReadouts().size());
+
+  // Set values for the rejected stats.
+  counter.inc();
+  gauge.inc();
+  textReadout.set("fortytwo");
+
+  // Ask for the rejected stats again by name.
+  Counter& counter2 = store_->counterFromString("c1");
```
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc:1228]: actually you can `ASSERT_EQ(&counter1, &counter2)`

File: test/common/stats/allocator_impl_test.cc [hunk 35]
```diff
@@ -125,6 +133,283 @@ TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
...
+  }
+  num_iterations = 0;
+  alloc_.forEachTextReadout(nullptr, [&num_iterations](Stats::TextReadout& text_readout) {
+    (void)text_readout;
+    ++num_iterations;
+  });
+  EXPECT_EQ(num_iterations, num_stats);
+}
+
+// Currently, if we ask for a stat from the  Allocator that has already been
```
Comment (jmarantz) [inline test/common/stats/allocator_impl_test.cc:342]: extra space after "the"

**Review by jmarantz — COMMENTED** [review]
looks great; just some minor comments.

**Review by jmarantz — APPROVED** [review]

**Review by mattklein123 — APPROVED** [review]
